packages/create-amplify/src/package-manager-controller/package_manager_controller_factory.ts

```diff
@@ -0,0 +1,157 @@
+import { existsSync as _existsSync } from 'fs';
+import { execa as _execa } from 'execa';
+import * as path from 'path';
+import {
+  type PackageManagerName,
+  type PackageManagers,
+} from './package_manager.js';
+import { logger } from '../logger.js';
+import { NpmPackageManagerController } from './npm_package_manager_controller.js';
+import { PnpmPackageManagerController } from './pnpm_package_manager_controller.js';
+import { YarnClassicPackageManagerController } from './yarn_classic_package_manager_controller.js';
+import { YarnModernPackageManagerController } from './yarn_modern_package_manager_controller.js';
+import {
+  NpmProjectInitializer,
+  PnpmProjectInitializer,
+  YarnClassicProjectInitializer,
+  YarnModernProjectInitializer,
+} from '../project-initializer/index.js';
+
+export type DependencyType = 'dev' | 'prod';
+
+/**
+ *
+ */
+export abstract class PackageManagerController {
+  abstract installDependencies: (
+    packageNames: string[],
+    type: DependencyType
+  ) => Promise<void>;
+}
+
+/**
+ * packageManagerControllerFactory
+ */
+export class PackageManagerControllerFactory {
+  private readonly existsSync = _existsSync;
+  packageManagerName: PackageManagerName;
+  private readonly packageManagers: PackageManagers = {
+    npm: {
+      name: 'npm',
+      executable: 'npm',
+      binaryRunner: 'npx',
+      installCommand: 'install',
+      lockFile: 'package-lock.json',
+      initDefault: ['init', '--yes'],
+    },
+    'yarn-classic': {
+      name: 'yarn-classic',
+      executable: 'yarn',
+      binaryRunner: 'yarn',
+      installCommand: 'add',
+      lockFile: 'yarn.lock',
+      initDefault: ['init', '--yes'],
+    },
+    'yarn-modern': {
+      name: 'yarn-modern',
+      executable: 'yarn',
+      binaryRunner: 'yarn',
+      installCommand: 'add',
+      lockFile: 'yarn.lock',
+      initDefault: ['init', '--yes'],
+    },
+    pnpm: {
+      name: 'pnpm',
+      executable: 'pnpm',
+      binaryRunner: 'pnpm',
+      installCommand: 'add',
+      lockFile: 'pnpm-lock.yaml',
+      initDefault: ['init'],
+    },
+  };
+
+  /**
+   * Check if a package.json file exists in projectRoot
+   */
+  protected packageJsonExists = (projectRoot: string): boolean => {
+    return this.existsSync(path.resolve(projectRoot, 'package.json'));
+  };
+
+  /**
+   * getPackageManager
+   */
+  getPackageManager() {
```

Comment:
since the package manager is the same for any given invocation of `create-amplify`, this should go in the constructor and store the package manager info as a private instance property

Reply:
done